src/compiler/checker.ts

```diff
@@ -16586,7 +16853,64 @@ namespace ts {
                 return resolveExternalModuleTypeByLiteral(<StringLiteral>node.arguments[0]);
             }
 
-            return getReturnTypeOfSignature(signature);
+            const returnType = getReturnTypeOfSignature(signature);
+            // Treat any call to the global 'Symbol' function that is part of a const variable or readonly property
+            // as a fresh unique symbol literal type.
+            if (returnType.flags & TypeFlags.ESSymbolLike && isSymbolOrSymbolForCall(node)) {
+                const parent = skipParentheses(node).parent;
+                const symbol = getSymbolOfNode(parent);
+                if (symbol) return getUniqueESSymbolTypeForSymbol(symbol);
+            }
+            return returnType;
+        }
+
+        function isSymbolOrSymbolForCall(node: Node) {
+            if (!isCallExpression(node)) return false;
+            let left = node.expression;
+            if (isPropertyAccessExpression(left) && left.name.escapedText === "for") {
+                left = left.expression;
+            }
+            if (!isIdentifier(left) || left.escapedText !== "Symbol") {
+                return false;
+            }
+
+            // make sure `Symbol` is the global symbol
+            const globalESSymbol = getGlobalESSymbolConstructorSymbol(/*reportErrors*/ false);
+            if (!globalESSymbol) {
+                return false;
+            }
+
+            return globalESSymbol === resolveName(left, "Symbol" as __String, SymbolFlags.Value, /*nameNotFoundMessage*/ undefined, /*nameArg*/ undefined, /*isUse*/ false);

```

Comment:
Rather than doing a syntactic lookup on the `for` and/or `Symbol` identifiers, then verifying that it is the same `Symbol`/`for` that is in the global scope, why not just check the property access with `checkPropertyAccessOrQualifiedName`, then verify that `getNodeLinks(node).resolvedSymbol === globalESSymbol`?

Reply:
This isn't substantially different from the `isCommonJsRequire` function beneath it, and this function is also used in `getTypeOfExpression` where we explicitly don't want to do a full type check.